### PR TITLE
Refactor load data

### PR DESCRIPTION
**Pull request by cwild-UoS**
Creating a draft PR for visibility on the refactor work.

At todays (18/10/22) meeting I'd like to discuss how/when we start merging the refactor code in

This refactor separates load data and identification into two functions. As well as removing `causal_test_case` from the constructor of the `CausalTestEngine`.

### CODE REVIEW

File: examples/covasim_/doubling_beta/causal_test_beta.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def doubling_beta_CATE_on_csv(observational_data_path: str, simulate_counterfact
 
     # Read in the observational data and perform identification
     past_execution_df = pd.read_csv(observational_data_path)
-    _, causal_test_engine = identification(observational_data_path)
+    _, causal_test_engine, causal_test_case = identification(observational_data_path)
```
Comment (AndrewC19) [inline examples/covasim_/doubling_beta/causal_test_beta.py:47]: I think this is an error -- `causal_test_case` should be passed as an argument to `identification`, right?

Comment (cwild-UoS) [reply examples/covasim_/doubling_beta/causal_test_beta.py:47]: Looking at it closer now, I think it is a logical function that would be clearer with a rename. Do you think its okay with how it is with the latest push?

**Comment by cwild-UoS** [comment]
After some discussion in a meeting yesterday (25/10/22), there are some more changes to come in this PR as a heads-up to anyone looking to review.

**Comment by cwild-UoS** [comment]
> After some discussion in a meeting yesterday (25/10/22), there are some more changes to come in this PR as a heads-up to anyone looking to review.

I think this PR is now good for review again @bobturneruk :)

**Review by bobturneruk — APPROVED** [review]
I'm a bit out of touch with the architecture of this package to comment on this aspect, but the code looks good and I think meets the objectives we've been discussing. A few specific comments which I'll leave for others to decide on whether changes are needed for this PR, or better for additional issues to be raised.

---

Error running the vaccinating_elderly example:

```
Traceback (most recent call last):
  File "C:\Users\bobturner\Documents\CausalTestingFramework\examples\covasim_\vaccinating_elderly\causal_test_vaccine.py", line 
226, in <module>
    test_results = experimental_causal_test_vaccinate_elderly(runs_per_test_per_config=30, verbose=True)
  File "C:\Users\bobturner\Documents\CausalTestingFramework\examples\covasim_\vaccinating_elderly\causal_test_vaccine.py", line 
84, in experimental_causal_test_vaccinate_elderly
    causal_test_engine.load_data(index_col=0)
AttributeError: 'CausalTestEngine' object has no attribute 'load_data'
```

File: causal_testing/testing/causal_test_engine.py [hunk 2]
```diff
@@ -29,25 +29,9 @@ class CausalTestEngine:
...
-        )
-        self.data_collector = data_collector
-        self.scenario_execution_data_df = pd.DataFrame()
-
-    def load_data(self, **kwargs):
-        """Load execution data corresponding to the causal test case into a pandas dataframe and return the minimal
-        adjustment set.
-
-        Data can be loaded in two ways:
+    Data is loaded as part of the "__init__" function
```
[the remark this answers is not in the document]
Comment (cwild-UoS) [reply causal_testing/testing/causal_test_engine.py:33]: Good spot, having a quick look around there are a couple of places with interesting formatting on sphinx. I'll make it an issue and make a separate PR to fix it / update it

File: causal_testing/json_front/json_class.py [hunk 3]
```diff
@@ -203,10 +203,10 @@ def _setup_test(self, causal_test_case: CausalTestCase, estimator: Estimator) ->
                 - estimation_model - Estimator instance for the test being run
         """
         data_collector = ObservationalDataCollector(self.modelling_scenario, self.data_path)
-        causal_test_engine = CausalTestEngine(causal_test_case, self.causal_specification, data_collector)
-        minimal_adjustment_set = causal_test_engine.load_data(index_col=0)
+        causal_test_engine = CausalTestEngine(self.causal_specification, data_collector, index_col=0)
```
[the remark this answers is not in the document]
Comment (cwild-UoS) [reply causal_testing/json_front/json_class.py:206]: I'll bring this up at the next meeting to discuss, I haven't dug as far as Pandas DFs yet

File: examples/covasim_/doubling_beta/causal_test_beta.py [hunk 4]
```diff
@@ -42,9 +42,9 @@ def doubling_beta_CATE_on_csv(observational_data_path: str, simulate_counterfact
     results_dict = {'association': {},
                     'causation': {}}
 
-    # Read in the observational data and perform identification
+    # Read in the observational data, perform identification, and setup the causal_test_engine
     past_execution_df = pd.read_csv(observational_data_path)
-    _, causal_test_engine = identification(observational_data_path)
+    _, causal_test_engine, causal_test_case = engine_setup(observational_data_path)
```
[the remark this answers is not in the document]
Comment (cwild-UoS) [reply examples/covasim_/doubling_beta/causal_test_beta.py:47]: `_` is used as a dummy or throwaway variable from my understanding. It's just assigns the return to a variable called `_` with the intent of not using it.

Linters like pylint I believe also accept this convention and will ignore variables called `_` not being used.

Whether its good practise or not I'm not sure on. The only alternative I can think of is returning the tuple whole and only unpacking the parts we want `causal_test_engine`. However its an extra line and won't be as clear imo?

Do you have any alternative suggestions that would be cleaner? It's a shame that Joe Heffer is on AL as he loves weighing in on meticulous python questions :)!

**Pull request merged**